portalnet/src/socket.rs

```diff
@@ -35,3 +45,82 @@ pub fn stun_for_external(local_socket_addr: &SocketAddr) -> Option<SocketAddr> {
         }
     }
 }
+
+pub fn is_local_addr(external_ip: Option<IpAddr>) -> bool {
+    if let Some(external_ip) = external_ip {
+        let network_interfaces = local_ip_address::list_afinet_netifas();
+
+        if let Ok(network_interfaces) = network_interfaces {
+            for (_, ip) in network_interfaces.iter() {
+                if *ip == external_ip {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+pub fn upnp_for_external(listen_addr: SocketAddr) -> Option<SocketAddr> {
+    info!("Connecting to UPnP gateway to map local address to external address");
+    let gateway = match igd_next::search_gateway(Default::default()) {
+        Ok(gateway) => gateway,
+        Err(ref err) => {
+            warn!(error = %err, "Error finding gateway");
+            return None;
+        }
+    };
+
+    let local_ip = match TcpStream::connect(gateway.addr) {
+        Ok(stream) => match stream.local_addr() {
+            Ok(addr) => addr.ip(),
+            Err(ref err) => {
+                warn!(error = %err, "Error finding local IP connected to gateway");
+                return None;
+            }
+        },
+        Err(ref err) => {
+            warn!(error = %err, "Error finding gateway");
+            return None;
+        }
+    };
+    let local_addr = SocketAddr::new(local_ip, listen_addr.port());
+
+    let external_ip = match gateway.get_external_ip() {
+        Ok(external_ip) => external_ip,
+        Err(ref err) => {
+            warn!(error = %err, "Error getting external IP");
+            return None;
+        }
+    };
+
+    match gateway.add_port(
+        igd_next::PortMappingProtocol::UDP,
+        listen_addr.port(),
+        local_addr,
+        UPNP_MAPPING_DURATION,
+        "new_port",
+    ) {
+        Ok(()) => {
+            thread::spawn(move || loop {
```

Comment:
This thread is used to renew the port periodically so that Trin can own the port as long as it is alive, while the port will be released in < 60 mins if Trin quits in any matter (crash or gracefully shutdown).

Reply:
Ahh, I see now, it makes sense, Let's use tokio to spawn the thread with tokio sleep:

```suggestion
            tokio::spawn(async move { loop {
                tokio::time::sleep(time::Duration::from_secs(UPNP_MAPPING_TIMEOUT)).await;
```

Reply:
A comment is added.

I did some study on `tokio::spawn`, and I believe that it may not be a good way because `gateway.add_port` into the UPnP library is a blocking IO, which may block other async tasks.  I tried to use `tokio::tasks::spawn_blocking`, but I encountered an issue that SIGINT (ctrl + c) cannot exit Trin gracefully.  As a result, I have to revert to the dedicated thread method for this purpose.

Please let me know if you have better ideas.

Reply:
Hmm, I didn't know that `gateway.add_port` is a blocking IO. In this case we have two options tokio spawn_blocking or a dedicated thread but because we expect this thread to run forever, I think `std::thread::spawn` is the way to go :+1: .